src/app/components/services/qhana-plugin.service.ts

```diff
@@ -0,0 +1,171 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { v4 as uuidv4 } from 'uuid';
+import { ImplementationDto } from 'api-nisq/models/implementation-dto';
+import { BehaviorSubject } from 'rxjs';
+
+interface MicroFrontendState {
+  href: string;
+  height: number;
+  initialized: boolean;
+}
+
+interface ImplementationItem {
+  name: string;
+  download: string;
+  version: number;
+  type: string;
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export class QhanaPluginService {
+  isPlugin: boolean = false;
+  implementationDtoSubject: BehaviorSubject<ImplementationDto[]> =
+    new BehaviorSubject([]);
+
+  qhanaFrontendState: MicroFrontendState = {
+    href: window.location.href,
+    height: 0,
+    initialized: false,
+  };
+  
+  algoUUIDs = new Map<string, string>();
+  implUUIDs = new Map<string, string>();
+
+  constructor(private http: HttpClient) {
+    this.isPlugin = window.top !== window.self;
+  }
+
+  /**
+   * Register the main message event listener on the current window.
+   */
+  registerMessageListener(): void {
+    // main event listener, delegates events to dedicated listeners
+    window.addEventListener('message', (event) => {
+      const data = event.data;
+      if (typeof data !== 'string') {
+        if (data != null && data.type === 'load-css') {
+          this.onLoadCssMessage(data);
+        } else if (data != null && data.type === 'implementations-response') {
+          this.onImplementationsResponse(data);
+        }
+      }
+    });
+  }
+
+  /**
+   * Handle css load messages that request the micro frontend to load additional css files.
+   *
+   * @param {{type: 'load-css', urls: string[]}} data
+   */
+  onLoadCssMessage(data: { urls: string[] }): void {
+    const head = document.querySelector('head');
+    data.urls.forEach((url) => {
+      const styleLink = document.createElement('link');
+      styleLink.href = url;
+      styleLink.rel = 'stylesheet';
+      head.appendChild(styleLink);
+    });
+    document.body.style.background = 'transparent';
+  }
+  
+  getUUID(impl: ImplementationItem, uuids: Map<string, string>): string {
```

Comment:
please use visibilities like private to hide internal methods. This helps to reduce the API surface and makes it easier to choose the correct method. Also document all public (and also all private) methods.